### PR TITLE
[SPARK-6352] [SQL] Add DirectParquetOutputCommitter

### PR DESCRIPTION
**Pull request by ypcat**
Add a DirectParquetOutputCommitter class that skips _temporary directory when saving to s3. To enable it, set hadoop configuration "spark.sql.parquet.output.committer.class" to "org.apache.spark.sql.parquet.DirectParquetOutputCommitter".

### CODE REVIEW

**Comment by ash211** [comment]
Jenkins this is ok to test

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by ypcat** [comment]
I change the design to allow more general usage. User can set "spark.sql.parquet.output.committer.class" to a class extending ParquetOutputFormat.
I still include the implementation of DirectParquetOutputCommitter, if it is unwanted, we can move it to be only used in test case.

**Comment by liancheng** [comment]
Hey @ypcat, sorry for the late review. If I understand this correctly, with `DirectParquetOutputCommitter`, all tasks writes to the final output directory directly. Then what if speculative tasks are scheduled? This may end up with two tasks trying to write to the same file simultaneously, right?

**Comment by liancheng** [comment]
At least we should state explicitly that speculative tasks should be disabled when using the direct output committer.

**Comment by JoshRosen** [comment]
@liancheng, if this is intended only for writing directly to S3, then I think we purposely want to avoid writing to `_temporary` because the move / rename of the temporary file to the final file is expensive and because S3 PUT operations are atomic.  If you have two tasks racing to write to the same file, then the last writer should win and we shouldn't have to worry about partially-written files.  As with DirectOutputCommitter, this is _not_ safe for use on filesystems where file creation / overwrite isn't atomic, such as HDFS, but that's not a use-case that we have to worry about here (I think this is only for use by advanced users who know that they're writing directly to S3).

**Comment by ypcat** [comment]
@liancheng this DirectParquetOutputCommitter is based on DirectOutputCommitter in this [thread](https://mail-archives.apache.org/mod_mbox/spark-user/201502.mbox/%3CCANGvG8qfmNPkb2vhL0-t9bi3d39b3ROhs6DhO+PODRRbzh-9Tg@mail.gmail.com%3E) and was intended to use on s3. I was not aware of the problem of speculation. I think we can either:
1. exclude DirectParquetOutputCommitter and let user implement their own like DirectOutputCommitter.
2. write down the consequences in documentation as @JoshRosen described.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableOperations.scala [hunk 1]
```diff
@@ -403,6 +405,26 @@ private[parquet] class AppendingParquetOutputFormat(offset: Int)
...
+        classOf[ParquetOutputCommitter], classOf[ParquetOutputCommitter])
+      val ctor = cls.getDeclaredConstructor(classOf[Path], classOf[TaskAttemptContext])
+      committer = ctor.newInstance(output, context).asInstanceOf[ParquetOutputCommitter]
+    }
+    committer
+  }
+
+  // FileOutputFormat.getOutputPath takes JobConf in hadoop-1 but JobContext in hadoop-2
+  private def getOutputPath(context: TaskAttemptContext): Path = {
+    context.getConfiguration().get("mapred.output.dir") match {
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableOperations.scala]: Should we check for `mapreduce.output.fileoutputformat.outputdir` instead?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableOperations.scala]: Oh I see, this is for Hadoop 1 compatibility.

**Comment by liancheng** [comment]
@JoshRosen @ypcat Thanks for the detailed explanation, I didn't realize S3 PUT operations are atomic at first. This LGTM. Merging to master.

@ypcat Thanks again for working on this!

**Comment by JoshRosen** [comment]
We can't change this now, but note that the commit message / PR description here is slightly out of date w.r.t. the actual changes being committed, since we now use the `spark.sql.parquet.output.committer.class` configuration instead of `spark.sql.parquet.useDirectParquetOutputCommitter`.

We may want to document this configuration at some point; if we do, I think the right place is the Parquet Configuration section in the SQL programming guide: https://spark.apache.org/docs/latest/sql-programming-guide.html#configuration

**Comment by JoshRosen** [comment]
Actually, on giving this a closer look I'm not sure whether this faithfully respects all of the possible Parquet configurations for controlling OutputCommitter behavior.  For instance, it looks like there's a `ParquetOutputFormat.ENABLE_JOB_SUMMARY` configuration that can be used to suppress the writing of the summary file (https://github.com/Parquet/parquet-mr/blame/fa8957d7939b59e8d391fa17000b34e865de015d/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputCommitter.java#L48).  I don't really know much about Parquet, so I can't really speak to whether this is a configuration that would be commonly used, but we should probably try to support it.

**Comment by ypcat** [comment]
@JoshRosen I updated PR description. It looks like the parquet code base I used was too old that it did not include the ParquetOutputFormat.ENABLE_JOB_SUMMARY thing. I will add support to this.

**Comment by JoshRosen** [comment]
It looks like this patch broke the Hadoop 1.x build because the test code uses `Configuration.unset`, which isn't present in Hadoop 1.x:

```
[info] Compiling 1 Scala source to /home/jenkins/workspace/Spark-Master-SBT/AMPLAB_JENKINS_BUILD_PROFILE/hadoop1.0/label/centos/examples/target/scala-2.10/classes...
[error] /home/jenkins/workspace/Spark-Master-SBT/AMPLAB_JENKINS_BUILD_PROFILE/hadoop1.0/label/centos/sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetIOSuite.scala:402: value unset is not a member of org.apache.hadoop.conf.Configuration
[error]       configuration.unset("spark.sql.parquet.output.committer.class")
[error]                     ^
[error] one error found
```

To unbreak the build, I'm going to revert this patch and re-open the JIRA.  Let's open a new PR to fix this.  At the same time, we can also address the ENABLE_JOB_SUMMARY flag.

**Comment by ypcat** [comment]
I cannot find a way to unset a config value in hadoop 1.x API. The closest thing is to set it to a default value, which I think should be fine in test code.
And I found I cannot add more commits to this PR since it is closed. Should we reopen it or use a new PR?

**Comment by ash211** [comment]
I think it's best to open a new PR

**Comment by ypcat** [comment]
New PR is https://github.com/apache/spark/pull/5525

**Pull request closed without merging**